src/replication/replication_source_db_copy.cpp

```diff
@@ -0,0 +1,643 @@
+    m_online_replication_start_pos = 0;
+
+    m_stream = NULL;
+    /* TODO[replication] : use file for replication copy stream */
+    m_stream_file = NULL;
+    m_transfer_sender = 0;
+    m_heap_extract_workers_pool = NULL;
+    m_state = NOT_STARTED;
+    m_is_stop = false;
+
+
+    m_stream = acquire_stream_for_copy ();
+    /* TODO : single global pool or a pool for each context ? */
+    m_heap_extract_workers_pool =
+	    cubthread::get_manager ()->create_worker_pool (EXTRACT_HEAP_WORKER_POOL_SIZE, EXTRACT_HEAP_WORKER_POOL_SIZE,
+		"replication_extract_heap_workers", NULL, 1, 1);
+  }
+
+  source_copy_context::~source_copy_context ()
+  {
+    cubthread::get_manager ()->destroy_worker_pool (m_heap_extract_workers_pool);
+    detach_stream_for_copy ();
+
+    m_stream_file = NULL;
+    delete m_transfer_sender;
+    m_transfer_sender = NULL;
+    m_state = NOT_STARTED;
+    m_error_cnt = 0;
+  }
+
+  void source_copy_context::pack_and_add_object (row_object *&obj)
+  {
+    if (obj->get_rec_cnt () == 0)
+      {
+	delete obj;
+	obj = NULL;
+	return;
+      }
+
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::ACTIVE);
+
+    stream_entry.add_packable_entry (obj);
+    /* once packed, the object is owned by the stream entry */
+    obj = NULL;
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_statement (const std::string &str)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::ACTIVE);
+    LOG_LSA null_LSA;
+    LSA_SET_NULL (&null_LSA);
+
+    sbr_repl_entry *sbr = new sbr_repl_entry (str.c_str (), "dba", "", null_LSA);
+    stream_entry.add_packable_entry (sbr);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_start_of_extract_heap (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::START_OF_EXTRACT_HEAP);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_end_of_extract_heap (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::END_OF_EXTRACT_HEAP);
+
+    stream_entry.pack ();
+  }
+
+  void source_copy_context::pack_and_add_end_of_copy (void)
+  {
+    stream_entry stream_entry (m_stream, MVCCID_FIRST, stream_entry_header::END_OF_REPLICATION_COPY);
+
+    stream_entry.pack ();
+  }
+
+  int source_copy_context::execute_and_transit_phase (copy_stage new_state)
+  {
+    int error = NO_ERROR;
+
+    assert (new_state != m_state);
+
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::execute_and_transit_phase "
+			      "curr_state:%d, new_state:%d", m_state, new_state);
+
+    std::unique_lock<std::mutex> ulock_state (m_state_mutex);
+
+    if ((int) m_state != ((int) new_state - 1))
+      {
+	return ER_FAILED;
+      }
+
+    m_state = new_state;
+
+    ulock_state.unlock ();
+
+    /* some of the new states require specific actions */
+    if (new_state == SCHEMA_APPLY_CLASSES_FINISHED)
```

Comment:
Is it safe to executed these specific actions here? What happens if this thread is preempted before calling pack_and_add_statement and thread waiting for state wakes up?

Should these actions should be executed by notified waiter?

Reply:
Extended life of mutex lock to include pack_and_add_statement as a comfort.

The "active control" of copy process is handed over from ddl_proxy to execute_db_copy thread when SCHEMA_CLASSES_LIST_FINISHED command is received (it is not included in this PR). There is no contention at this event (all actions are protected by mutex).

The pack_and_add_statement  actions cannot be executed by the waiter (the execute_db_copy thread) because the data is received from client side on the request thread; at most it may be added locally into the replication_source_context and send later to stream, but it would just increase the states of context and complexity.